pkg/proxy/node.go

```
@@ -109,3 +122,111 @@ func (n *NodeEligibleHandler) OnNodeDelete(node *v1.Node) { n.HealthServer.SyncN
+// and returns NodeManager after waiting for the node object to exist and have NodeIPs.
+// Note: NewNodeManager doesn't return any error if failed to retrieve NodeIPs.
+func NewNodeManager(ctx context.Context, client clientset.Interface, resyncInterval time.Duration, nodeName string) (*NodeManager, error) {
+	return newNodeManager(ctx, client, resyncInterval, nodeName, os.Exit, time.Second, 30*time.Second)
+}
+
+// newNodeManager implements NewNodeManager with configurable exit function, poll interval and timeout.
+func newNodeManager(ctx context.Context, client clientset.Interface, resyncInterval time.Duration, nodeName string, exitFunc func(int), pollInterval, pollTimeout time.Duration) (*NodeManager, error) {
+	// make an informer that selects for the given node
+	thisNodeInformerFactory := informers.NewSharedInformerFactoryWithOptions(client, resyncInterval,
+		informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", nodeName).String()
+		}))
+	nodeInformer := thisNodeInformerFactory.Core().V1().Nodes()
+	nodeLister := nodeInformer.Lister()
+
+	// initialize the informer and wait for cache sync
+	thisNodeInformerFactory.Start(wait.NeverStop)
+	if !cache.WaitForNamedCacheSync("node informer cache", ctx.Done(), nodeInformer.Informer().HasSynced) {
+		return nil, fmt.Errorf("can not sync node informer")
+	}
+
+	var node *v1.Node
+	var err error
+	var nodeIPs []net.IP
+
+	// wait for the node object to exist and have NodeIPs.
+	ctx, cancel := context.WithTimeout(ctx, pollTimeout)
+	defer cancel()
+	_ = wait.PollUntilContextCancel(ctx, pollInterval, true, func(context.Context) (bool, error) {
+		node, err = nodeLister.Get(nodeName)
+		if err != nil {
+			return false, nil
+		}
+
+		nodeIPs, err = utilnode.GetNodeHostIPs(node)
+		if err != nil {
+			return false, nil
+		}
+		return true, nil
+	})
```

you're discarding the error?

It seems to me like if kube-proxy can't get the node IPs, then it ought to fail... Can you do look through the old PRs that resulted in the current code to figure out why we made it keep going?

---

We do fallback to localhost (IPv4 and IPv6) when NodeIPs are not detected
https://github.com/kubernetes/kubernetes/blob/b98b86bd6abb24e3d2451a885354f09cd7af22a2/cmd/kube-proxy/app/server.go#L636-L649

I'll go through the old PRs.

---

Historically `--bind-address` was directly used as NodeIP for primary IP Family detection, and https://github.com/kubernetes/kubernetes/commit/56df70b639c5b5bf53076a3483a2e40b8f60c49a (https://github.com/kubernetes/kubernetes/pull/91725) added `detectNodeIP` which fetched NodeIP from Node object if the `--bind-address` was unspecified and fall backed to localhost on failure.

https://github.com/kubernetes/kubernetes/commit/0df4a69f5c90990d47d1337e5fc99a764ff99474 added the polling to wait for NodeIPs to be available with Node object, but never returned any error on poll timeout.

---

So, I guess the reason we ignore errors is because the node IP is _mostly_ not actually used, other than for deciding the primary IP family, so it seemed weird to make a `GetNodeHostIPs` failure be fatal... But at the same time, it really _should_ be fatal...

Either way, I dislike the fact that we completely throw the error away within NodeManager. If we're going to ignore the error, it ought to be done at the `cmd/kube-proxy` level.

And actually, the current code doesn't work anyway; kubelet always sets _some_ node IPs on the node when it first creates the Node object, so the only way we could get to the end of `newNodeManager` without having found node IPs would be if we didn't even find a Node, in which case we'd end up setting `n.node` to `nil` and then `n.NodeIPs()` would crash inside `utilnode.GetNodeHostIPs()`.

At this point we have basically normalized the fact that kube-proxy will exit and expect to be restarted in some circumstances. We should just treat the lack of a Node object as a fatal error and exit.